torchvision/prototype/features/_dataset_wrapper.py

```diff
@@ -0,0 +1,388 @@
+            **(dtypes or dict()),
+        }
+        wrappers_fn = cls._wrappers_fns[type(dataset)]
+        wrappers = wrappers_fn(dataset, keep_pil_image, bounding_box_format, dtypes)
+        return cls(dataset, wrappers)
+
+
+def identity_wrapper(obj: T) -> T:
+    return obj
+
+
+def generic_feature_wrapper(data: Any) -> features.GenericFeature:
+    return features.GenericFeature(data)
+
+
+def wrap_image(image: PIL.Image.Image, *, dtype: Optional[torch.dtype]) -> features.Image:
+    image = F.pil_to_tensor(image)
+    if dtype is not None:
+        image = F.convert_image_dtype(image, dtype=dtype)
+    return features.Image(image)
+
+
+def make_image_wrapper(
+    *, keep_pil_image: bool, dtypes: Dict[Type[features._Feature], Optional[torch.dtype]]
+) -> Callable[[PIL.Image.Image], Union[PIL.Image.Image, features.Image]]:
+    if keep_pil_image:
+        return identity_wrapper
+
+    return functools.partial(wrap_image, dtype=dtypes[features.Image])
+
+
+def wrap_label(label: Any, *, categories: Optional[Sequence[str]], dtype: Optional[torch.dtype]) -> features.Label:
+    return features.Label(label, categories=categories, dtype=dtype)
+
+
+def make_label_wrapper(
+    *, categories: Optional[Sequence[str]], dtypes: Dict[Type[features._Feature], Optional[torch.dtype]]
+) -> Callable[[Any], features.Label]:
+    return functools.partial(wrap_label, categories=categories, dtype=dtypes[features.Label])
+
+
+def wrap_segmentation_mask(segmentation_mask: PIL.Image.Image, *, dtype: Optional[torch.dtype]) -> features.Mask:
+    assert isinstance(segmentation_mask, PIL.Image.Image)
+
+    segmentation_mask = F.pil_to_tensor(segmentation_mask)
+    if dtype is not None:
+        segmentation_mask = F.convert_image_dtype(segmentation_mask, dtype=dtype)
+    return features.Mask(segmentation_mask.squeeze(0))
+
+
+def make_segmentation_mask_wrapper(
+    *, dtypes: Dict[Type[features._Feature], Optional[torch.dtype]]
+) -> Callable[[Any], features.Mask]:
+    return functools.partial(wrap_segmentation_mask, dtype=dtypes[features.Mask])
+
+
+CATEGORIES_GETTER = defaultdict(
+    lambda: lambda dataset: None,
+    {
+        datasets.Caltech256: lambda dataset: [name.rsplit(".", 1)[1] for name in dataset.categories],
+        datasets.CIFAR10: lambda dataset: dataset.classes,
+        datasets.CIFAR100: lambda dataset: dataset.classes,
+        datasets.FashionMNIST: lambda dataset: dataset.classes,
+        datasets.ImageNet: lambda dataset: [", ".join(names) for names in dataset.classes],
+    },
+)
+
+
+def classification_wrappers(
+    dataset: datasets.VisionDataset,
+    keep_pil_image: bool,
+    bounding_box_format: Optional[features.BoundingBoxFormat],
+    dtypes: Dict[Type[features._Feature], Optional[torch.dtype]],
+) -> Any:
+    return (
+        make_image_wrapper(keep_pil_image=keep_pil_image, dtypes=dtypes),
+        make_label_wrapper(categories=CATEGORIES_GETTER[type(dataset)](dataset), dtypes=dtypes),
+    )
```

Comment:
I'm wondering how much of this machinery is needed. At its core all we need for classification datasets is something like this (ignoring dtype and keep_pil_image which hopefully don't add too much complexity)


```py
class ClassificationWrapper(VisionDataset):
    def __getitem__(self, idx):
        image, target = super().__getitem__(idx)
        return Image(image), Target(target)
```


Granted I'm cheating a bit and this doesn't support everything the PR supports. But I'm wondering where all the extra code / logic in the PR comes from. We have to jump through a bunch of builders, classes, methods, helpers and partials to do something that should seem pretty basic.

Is this so that we can handle more complex cases like segmentation similarly? Are there opportunities for simplification?

Reply:
> Is this so that we can handle more complex cases like segmentation similarly? Are there opportunities for simplification?

Most likely. Let me have another look now that I haven't done so in a few months. I tend to overcomplicate stuff in the development phase. Will ping you when I have addressed the comments above and made it simpler.